[native-image] Vanilla project does not work with GraalVM 19.3.0

Micronaut 1.2.8 (and 1.3.0.M2)
GraalVM 19.3.0.

```
implementation 'io.micronaut.configuration:micronaut-flyway'
```

Vanilla project created. 

```
$ native-image --no-server -cp build/libs/my-service-0.1-all.jar
```

```
$ ./my-service
13:12:23.677 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties] could not be loaded: Error instantiating bean of type [io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties]: null
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties] could not be loaded: Error instantiating bean of type [io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties]: null
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1264)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:236)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2446)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:200)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:187)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:69)
	at com.pencepay.webhook.Application.main(Application.kt:14)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties]: null
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1626)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2307)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:1888)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1262)
	... 6 common frames omitted
Caused by: java.lang.ExceptionInInitializerError: null
	at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:290)
	at java.lang.Class.ensureInitialized(DynamicHub.java:475)
	at org.flywaydb.core.internal.util.FeatureDetector.isSlf4jAvailable(FeatureDetector.java:96)
	at org.flywaydb.core.internal.logging.LogCreatorFactory.getLogCreator(LogCreatorFactory.java:39)
	at org.flywaydb.core.api.logging.LogFactory.getLog(LogFactory.java:78)
	at org.flywaydb.core.internal.util.FeatureDetector.<clinit>(FeatureDetector.java:25)
	at com.oracle.svm.core.hub.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:350)
	at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:270)
	at java.lang.Class.ensureInitialized(DynamicHub.java:475)
	at org.flywaydb.core.internal.logging.LogCreatorFactory.getLogCreator(LogCreatorFactory.java:35)
	at org.flywaydb.core.api.logging.LogFactory.getLog(LogFactory.java:78)
	at org.flywaydb.core.api.configuration.ClassicConfiguration.<clinit>(ClassicConfiguration.java:57)
	at com.oracle.svm.core.hub.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:350)
	at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:270)
	at java.lang.Class.ensureInitialized(DynamicHub.java:475)
	at org.flywaydb.core.api.configuration.FluentConfiguration.<init>(FluentConfiguration.java:46)
	at io.micronaut.configuration.dbmigration.flyway.FlywayConfigurationProperties.<init>(FlywayConfigurationProperties.java:48)
	at io.micronaut.configuration.dbmigration.flyway.$FlywayConfigurationPropertiesDefinition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:117)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:109)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1598)
	... 9 common frames omitted
Caused by: org.flywaydb.core.api.FlywayException: Unable to instantiate class org.flywaydb.core.internal.logging.javautil.JavaUtilLogCreator : org.flywaydb.core.internal.logging.javautil.JavaUtilLogCreator
	at org.flywaydb.core.internal.util.ClassUtils.instantiate(ClassUtils.java:63)
	at org.flywaydb.core.internal.logging.LogCreatorFactory.getLogCreator(LogCreatorFactory.java:46)
	at org.flywaydb.core.api.logging.LogFactory.getLog(LogFactory.java:78)
	at org.flywaydb.core.internal.util.ClassUtils.<clinit>(ClassUtils.java:39)
	at com.oracle.svm.core.hub.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:350)
	at com.oracle.svm.core.hub.ClassInitializationInfo.initialize(ClassInitializationInfo.java:270)
	... 29 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.flywaydb.core.internal.logging.javautil.JavaUtilLogCreator
	at com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:60)
	at java.lang.Class.forName(DynamicHub.java:1197)
	at org.flywaydb.core.internal.util.ClassUtils.instantiate(ClassUtils.java:61)
	... 34 common frames omitted
```

Do you know how to fix this? Problem seems to be in `Caused by: java.lang.ClassNotFoundException: org.flywaydb.core.internal.logging.javautil.JavaUtilLogCreator`

Adding it with `@TypeHint` to Application did not help.

So this can be fixed using something like https://github.com/barryku/karaf-flyway-demo/pull/1/commits/f04d3a95a491f7202c02a537ee740d89d71a844f  and then adding `LogFactory.setLogCreator(Slf4jLogCreator())` before Flyway is initialized (this should be added to micronaut-flyway), like in that commit.

But, then there is a new problem after a recompiled native executable is started: 

```
20:14:45.214 [pool-2-thread-1] INFO  i.m.c.d.f.AbstractFlywayMigration - Running migrations for database with qualifier [default]

....
Caused by: org.flywaydb.core.api.FlywayException: Unable to obtain inputstream for resource: org/flywaydb/core/internal/version.txt
```

This has something to do with which ClassLoader is used to retrieve version.txt

In any case micronaut-flyway does not work with GraalVM. I suggest taking all the solutions from Quarkus Flyway extension, as they have solved all the problems already with flyway running native.

--- a/flyway/build.gradle
+++ b/flyway/build.gradle
@@ -1,8 +1,10 @@
 dependencies {
     annotationProcessor "io.micronaut:micronaut-inject-java"
+    annotationProcessor "io.micronaut:micronaut-graal"
     annotationProcessor "io.micronaut.docs:micronaut-docs-asciidoc-config-props:$micronautDocsVersion"
 
     compileOnly "io.micronaut:micronaut-inject-java"
+    compileOnly "org.graalvm.nativeimage:svm"
     compileOnly "org.grails:grails-datastore-gorm-hibernate5:$gormHibernateVersion"
 
     api "org.flywaydb:flyway-core:$flywayVersion"

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/FeatureDetectorSubstitutions.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/FeatureDetectorSubstitutions.java
@@ -17,15 +17,20 @@ package io.micronaut.flyway.graalvm;
 
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
+import io.micronaut.core.annotation.Internal;
 
 /**
  * FeatureDetector substitution.
  *
- * Forked from the Quarkus implementation.
+ * Forked from the Quarkus: https://github.com/quarkusio/quarkus/blob/38280358d8cf04b91dad7ef5a9d463823f8ae675/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/graal/FeatureDetectorSubstitutions.java
+ *
+ * @author Iván López
+ * @since 2.0.0
  */
 @Substitute
+@Internal
 @TargetClass(className = "org.flywaydb.core.internal.util.FeatureDetector")
-public final class FeatureDetectorSubstitutions {
+final class FeatureDetectorSubstitutions {
 
     @Substitute
     public FeatureDetectorSubstitutions(ClassLoader classLoader) {

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/FeatureDetectorSubstitutions.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/FeatureDetectorSubstitutions.java
@@ -1,0 +1,64 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.flyway.graalvm;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+/**
+ * FeatureDetector substitution.
+ *
+ * Forked from the Quarkus implementation.
+ */
+@Substitute
+@TargetClass(className = "org.flywaydb.core.internal.util.FeatureDetector")
+public final class FeatureDetectorSubstitutions {
+
+    @Substitute
+    public FeatureDetectorSubstitutions(ClassLoader classLoader) {
+    }
+
+    @Substitute
+    public boolean isApacheCommonsLoggingAvailable() {
+        return false;
+    }
+
+    @Substitute
+    public boolean isSlf4jAvailable() {
+        return true;
+    }
+
+    @Substitute
+    public boolean isJBossVFSv2Available() {
+        return false;
+    }
+
+    @Substitute
+    public boolean isJBossVFSv3Available() {
+        return false;
+    }
+
+    @Substitute
+    public boolean isOsgiFrameworkAvailable() {
+        return false;
+    }
+
+    @Substitute
+    public boolean isAndroidAvailable() {
+        return false;
+    }
+
+}

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/FlywayFeature.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/FlywayFeature.java
@@ -42,7 +42,7 @@ import java.util.stream.Stream;
 /**
  * A GraalVM feature that calculates the migration to apply at compile time.
  *
- * Forked from the Quarkus implementation.
+ * Forked from Quarkus: https://github.com/quarkusio/quarkus/blob/7a5efed2a97d88656484b431b472210e2bb7d2f3/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java
  *
  * @author Iván López
  * @since 2.0.0

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/FlywayFeature.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/FlywayFeature.java
@@ -1,0 +1,130 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.flyway.graalvm;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import io.micronaut.core.annotation.Internal;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A GraalVM feature that calculates the migration to apply at compile time.
+ *
+ * Forked from the Quarkus implementation.
+ *
+ * @author Iván López
+ * @since 2.0.0
+ */
+@Internal
+@AutomaticFeature
+final class FlywayFeature implements Feature {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlywayFeature.class);
+
+    private static final String CLASSPATH_APPLICATION_MIGRATIONS_PROTOCOL = "classpath";
+    private static final String JAR_APPLICATION_MIGRATIONS_PROTOCOL = "jar";
+    private static final String FILE_APPLICATION_MIGRATIONS_PROTOCOL = "file";
+
+    private static final String FLYWAY_LOCATIONS = "flyway.locations";
+    private static final String DEFAULT_FLYWAY_LOCATIONS = "classpath:db/migration";
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        List<String> locations = Stream
+                .of(System.getProperty(FLYWAY_LOCATIONS, DEFAULT_FLYWAY_LOCATIONS).split(","))
+                .collect(Collectors.toList());
+
+        try {
+            List<String> migrations = discoverApplicationMigrations(locations);
+            MicronautPathLocationScanner.setApplicationMigrationFiles(migrations);
+        } catch (IOException | URISyntaxException e) {
+            LOG.error("There was an error discovering the Flyway migrations: {}", e.getMessage());
+        }
+    }
+
+    private List<String> discoverApplicationMigrations(List<String> locations) throws IOException, URISyntaxException {
+        try {
+            List<String> applicationMigrationResources = new ArrayList<>();
+            // Locations can be a comma separated list
+            for (String location : locations) {
+                // Strip any 'classpath:' protocol prefixes because they are assumed
+                // but not recognized by ClassLoader.getResources()
+                if (location != null && location.startsWith(CLASSPATH_APPLICATION_MIGRATIONS_PROTOCOL + ':')) {
+                    location = location.substring(CLASSPATH_APPLICATION_MIGRATIONS_PROTOCOL.length() + 1);
+                }
+                Enumeration<URL> migrations = Thread.currentThread().getContextClassLoader().getResources(location);
+                while (migrations.hasMoreElements()) {
+                    URL path = migrations.nextElement();
+                    LOG.debug("Adding application migrations in path '{}' using protocol '{}'", path.getPath(), path.getProtocol());
+                    final Set<String> applicationMigrations;
+                    if (JAR_APPLICATION_MIGRATIONS_PROTOCOL.equals(path.getProtocol())) {
+                        try (FileSystem fileSystem = initFileSystem(path.toURI())) {
+                            applicationMigrations = getApplicationMigrationsFromPath(location, path);
+                        }
+                    } else if (FILE_APPLICATION_MIGRATIONS_PROTOCOL.equals(path.getProtocol())) {
+                        applicationMigrations = getApplicationMigrationsFromPath(location, path);
+                    } else {
+                        LOG.warn("Unsupported URL protocol '{}' for path '{}'. Migration files will not be discovered.", path.getProtocol(), path.getPath());
+                        applicationMigrations = null;
+                    }
+                    if (applicationMigrations != null) {
+                        applicationMigrationResources.addAll(applicationMigrations);
+                    }
+                }
+            }
+            return applicationMigrationResources;
+        } catch (IOException | URISyntaxException e) {
+            throw e;
+        }
+    }
+
+    private Set<String> getApplicationMigrationsFromPath(final String location, final URL path)
+            throws IOException, URISyntaxException {
+        try (Stream<Path> pathStream = Files.walk(Paths.get(path.toURI()))) {
+            return pathStream.filter(Files::isRegularFile)
+                    .map(it -> Paths.get(location, it.getFileName().toString()).toString())
+                    // we don't want windows paths here since the paths are going to be used as classpath paths anyway
+                    .map(it -> it.replace('\\', '/'))
+                    .peek(it -> LOG.trace("Discovered path: {}", it))
+                    .collect(Collectors.toSet());
+        }
+    }
+
+    private FileSystem initFileSystem(final URI uri) throws IOException {
+        final Map<String, String> env = new HashMap<>();
+        env.put("create", "true");
+        return FileSystems.newFileSystem(uri, env);
+    }
+}

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/MicronautPathLocationScanner.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/MicronautPathLocationScanner.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.flyway.graalvm;
+
+import io.micronaut.context.condition.Condition;
+import org.flywaydb.core.api.Location;
+import org.flywaydb.core.internal.resource.LoadableResource;
+import org.flywaydb.core.internal.resource.classpath.ClassPathResource;
+import org.flywaydb.core.internal.scanner.classpath.ResourceAndClassScanner;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This class is used in order to prevent Flyway from doing classpath scanning which is both slow and won't work in
+ * native mode.
+ *
+ * Forked from the Quarkus implementation.
+ */
+@SuppressWarnings("rawtypes")
+public final class MicronautPathLocationScanner implements ResourceAndClassScanner {
+    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(Condition.class);
+
+    private static final String LOCATION_SEPARATOR = "/";
+    private static List<String> applicationMigrationFiles;
+
+    private final Collection<LoadableResource> scannedResources;
+
+    public MicronautPathLocationScanner(Collection<Location> locations) {
+        this.scannedResources = new ArrayList<>();
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+
+        for (String migrationFile : applicationMigrationFiles) {
+            if (canHandleMigrationFile(locations, migrationFile)) {
+                LOG.debug("Loading %{}", migrationFile);
+                scannedResources.add(new ClassPathResource(null, migrationFile, classLoader, StandardCharsets.UTF_8));
+            }
+        }
+    }
+
+    @Override
+    public Collection<LoadableResource> scanForResources() {
+        return scannedResources;
+    }
+
+    private boolean canHandleMigrationFile(Collection<Location> locations, String migrationFile) {
+        for (Location location : locations) {
+            String locationPath = location.getPath();
+            if (!locationPath.endsWith(LOCATION_SEPARATOR)) {
+                locationPath += "/";
+            }
+
+            if (migrationFile.startsWith(locationPath)) {
+                return true;
+            } else {
+                LOG.debug("Migration file '{}' will be ignored because it does not start with '{}'", migrationFile, locationPath);
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public Collection<Class<?>> scanForClasses() {
+        // Classes are not supported in native mode
+        return Collections.emptyList();
+    }
+
+    public static void setApplicationMigrationFiles(List<String> applicationMigrationFiles) {
+        MicronautPathLocationScanner.applicationMigrationFiles = applicationMigrationFiles;
+    }
+}

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/MicronautPathLocationScanner.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/MicronautPathLocationScanner.java
@@ -16,6 +16,7 @@
 package io.micronaut.flyway.graalvm;
 
 import io.micronaut.context.condition.Condition;
+import io.micronaut.core.annotation.Internal;
 import org.flywaydb.core.api.Location;
 import org.flywaydb.core.internal.resource.LoadableResource;
 import org.flywaydb.core.internal.resource.classpath.ClassPathResource;
@@ -32,10 +33,14 @@ import java.util.List;
  * This class is used in order to prevent Flyway from doing classpath scanning which is both slow and won't work in
  * native mode.
  *
- * Forked from the Quarkus implementation.
+ * Forked from the Quarkus: https://github.com/quarkusio/quarkus/blob/7a5efed2a97d88656484b431b472210e2bb7d2f3/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/QuarkusPathLocationScanner.java
+ *
+ * @author Iván López
+ * @since 2.0.0
  */
 @SuppressWarnings("rawtypes")
-public final class MicronautPathLocationScanner implements ResourceAndClassScanner {
+@Internal
+final class MicronautPathLocationScanner implements ResourceAndClassScanner {
     private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(Condition.class);
 
     private static final String LOCATION_SEPARATOR = "/";
@@ -60,6 +65,16 @@ public final class MicronautPathLocationScanner implements ResourceAndClassScann
         return scannedResources;
     }
 
+    @Override
+    public Collection<Class<?>> scanForClasses() {
+        // Classes are not supported in native mode
+        return Collections.emptyList();
+    }
+
+    public static void setApplicationMigrationFiles(List<String> applicationMigrationFiles) {
+        MicronautPathLocationScanner.applicationMigrationFiles = applicationMigrationFiles;
+    }
+
     private boolean canHandleMigrationFile(Collection<Location> locations, String migrationFile) {
         for (Location location : locations) {
             String locationPath = location.getPath();
@@ -75,15 +90,5 @@ public final class MicronautPathLocationScanner implements ResourceAndClassScann
         }
 
         return false;
-    }
-
-    @Override
-    public Collection<Class<?>> scanForClasses() {
-        // Classes are not supported in native mode
-        return Collections.emptyList();
-    }
-
-    public static void setApplicationMigrationFiles(List<String> applicationMigrationFiles) {
-        MicronautPathLocationScanner.applicationMigrationFiles = applicationMigrationFiles;
     }
 }

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/ScannerSubstitutions.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/ScannerSubstitutions.java
@@ -18,6 +18,7 @@ package io.micronaut.flyway.graalvm;
 import com.oracle.svm.core.annotate.Alias;
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
+import io.micronaut.core.annotation.Internal;
 import org.flywaydb.core.api.Location;
 import org.flywaydb.core.internal.resource.LoadableResource;
 import org.flywaydb.core.internal.scanner.LocationScannerCache;
@@ -32,10 +33,14 @@ import java.util.List;
 /**
  * This substitution replaces the Flyway dynamic scanners with a fixed path scanner in native mode.
  *
- * Forked from the Quarkus implementation.
+ * Forked from the Quarkus: https://github.com/quarkusio/quarkus/blob/2c99a30985e7cae42933b949ecd2ee82d546c4aa/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/graal/ScannerSubstitutions.java
+ *
+ * @author Iván López
+ * @since 2.0.0
  */
+@Internal
 @TargetClass(className = "org.flywaydb.core.internal.scanner.Scanner")
-public final class ScannerSubstitutions {
+final class ScannerSubstitutions {
 
     @Alias
     private List<LoadableResource> resources = new ArrayList<>();

--- a/flyway/src/main/java/io/micronaut/flyway/graalvm/ScannerSubstitutions.java
+++ b/flyway/src/main/java/io/micronaut/flyway/graalvm/ScannerSubstitutions.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.flyway.graalvm;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.flywaydb.core.api.Location;
+import org.flywaydb.core.internal.resource.LoadableResource;
+import org.flywaydb.core.internal.scanner.LocationScannerCache;
+import org.flywaydb.core.internal.scanner.ResourceNameCache;
+import org.flywaydb.core.internal.scanner.classpath.ResourceAndClassScanner;
+
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * This substitution replaces the Flyway dynamic scanners with a fixed path scanner in native mode.
+ *
+ * Forked from the Quarkus implementation.
+ */
+@TargetClass(className = "org.flywaydb.core.internal.scanner.Scanner")
+public final class ScannerSubstitutions {
+
+    @Alias
+    private List<LoadableResource> resources = new ArrayList<>();
+
+    @Alias
+    private List<Class<?>> classes = new ArrayList<>();
+
+    /**
+     * Creates only {@link MicronautPathLocationScanner} instances.
+     * Replaces the original method that tries to detect migrations using reflection techniques that are not allowed
+     * in native mode.
+     *
+     * @see org.flywaydb.core.internal.scanner.Scanner#Scanner(Class, Collection, ClassLoader, Charset, ResourceNameCache,
+     * LocationScannerCache)
+     */
+    @SuppressWarnings("checkstyle:javadocmethod")
+    @Substitute
+    public ScannerSubstitutions(Class<?> implementedInterface, Collection<Location> locations, ClassLoader classLoader,
+                                Charset encoding, ResourceNameCache resourceNameCache, LocationScannerCache locationScannerCache) {
+        ResourceAndClassScanner scanner = new MicronautPathLocationScanner(locations);
+
+        Collection resources = scanner.scanForResources();
+        this.resources.addAll(resources);
+
+        Collection scanForClasses = scanner.scanForClasses();
+        classes.addAll(scanForClasses);
+    }
+
+}

--- a/flyway/src/main/resources/META-INF/native-image/io.micronaut.flyway/flyway/native-image.properties
+++ b/flyway/src/main/resources/META-INF/native-image/io.micronaut.flyway/flyway/native-image.properties
@@ -1,0 +1,2 @@
+Args = -H:IncludeResources=org/flywaydb/core/internal/version.txt \
+       --initialize-at-run-time=org.flywaydb.core.internal.util.FeatureDetector

--- a/flyway/src/main/resources/META-INF/native-image/io.micronaut.flyway/flyway/reflect-config.json
+++ b/flyway/src/main/resources/META-INF/native-image/io.micronaut.flyway/flyway/reflect-config.json
@@ -1,0 +1,6 @@
+[
+  {
+    "name":"org.flywaydb.core.internal.logging.slf4j.Slf4jLogCreator",
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  }
+]

--- a/src/main/docs/guide/graalvm.adoc
+++ b/src/main/docs/guide/graalvm.adoc
@@ -2,7 +2,7 @@
 Micronaut Flyway is compatible with https://www.graalvm.org/[GraalVM] so it is possible
 to create native images and run the migrations during application startup.
 
-The list of migrations to run is precomputed during native image build time and by default to process looks for migrations
+The list of migrations to run is precomputed during native image build time and by default Micronaut looks for migrations
 in the Flyway default directory `src/main/resources/db/migration`. If you want to change that directory or add more directories
 to look for migrations you need to add an additional parameter to `native-image` command.
 

--- a/src/main/docs/guide/graalvm.adoc
+++ b/src/main/docs/guide/graalvm.adoc
@@ -1,0 +1,34 @@
+
+Micronaut Flyway is compatible with https://www.graalvm.org/[GraalVM] so it is possible
+to create native images and run the migrations during application startup.
+
+The list of migrations to run is precomputed during native image build time and by default to process looks for migrations
+in the Flyway default directory `src/main/resources/db/migration`. If you want to change that directory or add more directories
+to look for migrations you need to add an additional parameter to `native-image` command.
+
+For example, if you have the following configuration:
+
+[source,yaml]
+----
+flyway:
+  datasources:
+    default:
+      locations:
+        - classpath:databasemigrations # <1>
+        - classpath:other # <1>
+----
+<1> Flyway will look for migrations in `src/main/resources/databasemigrations` and `src/main/resources/other`.
+
+You need to use the parameter `flyway.locations` with a comma separated list of directories to look for:
+
+[source,bash]
+----
+native-image --no-server --no-fallback \
+    --class-path build/libs/micronaut-flyway-graal-*-all.jar \
+    -Dflyway.locations='classpath:databasemigrations,classpath:other' # <1>
+----
+<1> Define the locations to look for migrations.
+
+
+NOTE: See the section on https://docs.micronaut.io/latest/guide/index.html#graal[GraalVM] in the user guide for more
+information.

--- a/src/main/docs/guide/toc.yml
+++ b/src/main/docs/guide/toc.yml
@@ -6,6 +6,7 @@ configuration:
   jpa: JPA/Hibernate
   gorm: GORM
   additionalConfig: Additional configuration
+graalvm: GraalVM support
 events: Events
 endpoint: Endpoint
 repository: Repository

--- a/src/main/docs/guide/whatsNew.adoc
+++ b/src/main/docs/guide/whatsNew.adoc
@@ -3,6 +3,7 @@ This section will outline any breaking changes between major or milestone releas
 == 2.0.0.M2
 
 * By default look for migrations in Flyway default location: `src/main/resources/db/migration`.
+* Add support for GraalVM.
 
 == 2.0.0.M1
 

